File: backend/src/grimoire/types/hud.py
"""Types for the Scene HUD aggregator.

Widget descriptors a mechanics module can declare in its ``manifest.yaml``
under ``hud_widgets``, plus the snapshot/aggregate shapes the HUD service
returns. The HUD owns no state of its own — these structures describe
where to read and edit, and how to render, but the data lives with the
canonical owner module.

See ``docs/superpowers/specs/2026-05-19-scene-hud-design.md``.
"""

from __future__ import annotations

from enum import StrEnum
from typing import Any

from pydantic import BaseModel, Field


class RenderHint(StrEnum):
    """How the frontend should lay out a widget."""

    ROW = "row"
    BLOCK = "block"
    CHIP_LIST = "chip-list"
    BANNER = "banner"
    COMPOSITE = "composite"


class WidgetScope(StrEnum):
    """What entity the widget applies to."""

    CAMPAIGN = "campaign"
    SCENE = "scene"
    PC = "pc"
    PRESENT_NPC = "present_npc"


class WidgetEditKind(StrEnum):
    INLINE_TEXT = "inline-text"
    PICKER = "picker"
    SLIDER = "slider"
    ENUM = "enum"
    COMPOSITE = "composite"


class WidgetRead(BaseModel):
    endpoint: str
    poll_interval_s: int | None = None


class WidgetEdit(BaseModel):
    kind: str
    endpoint: str
    schema_ref: str | None = None


class HudWidget(BaseModel):
    """A single dashboard widget descriptor."""

    id: str
    title: str
    scope: WidgetScope = WidgetScope.CAMPAIGN
    visible_when: str | None = None
    # Validated at aggregation time — unknown hints log once and fall back
    # to BLOCK so old frontends keep working with new modules.
    render_hint: str = RenderHint.BLOCK.value
    read: WidgetRead
    edit: WidgetEdit | None = None
    refresh_on: list[str] = Field(default_factory=list)
    stale_threshold_s: int | None = None
    owner_module: str | None = None


class WidgetStatus(StrEnum):
    OK = "ok"
    ERROR = "error"
    TIMEOUT = "timeout"
    HIDDEN = "hidden"


class WidgetSnapshot(BaseModel):
    """A single widget's resolved state from one aggregate call."""

    id: str
    status: WidgetStatus = WidgetStatus.OK
    data: Any | None = None
    error: str | None = None
    stale: bool = False
    title: str | None = None
    render_hint: str | None = None


class AggregateResult(BaseModel):
    """Full HUD payload returned by ``GET /campaigns/{id}/hud``."""

    campaign_id: str
    scene_id: str | None = None
    generated_at: str
    widgets: list[WidgetSnapshot] = Field(default_factory=list)


__all__ = [
    "AggregateResult",
    "HudWidget",
    "RenderHint",
    "WidgetEdit",
    "WidgetEditKind",
    "WidgetRead",
    "WidgetScope",
    "WidgetSnapshot",
    "WidgetStatus",
]

File: frontend/src/api/hud.ts
/**
 * Scene HUD client. Mirrors the Pydantic types in
 * ``backend/src/grimoire/types/hud.py`` and the routes in
 * ``backend/src/grimoire/api/hud.py``.
 */

import { api } from "./client";

export type WidgetStatus = "ok" | "error" | "timeout" | "hidden";

export type RenderHint = "row" | "block" | "chip-list" | "banner" | "composite";

export type WidgetScope = "campaign" | "scene" | "pc" | "present_npc";

export interface WidgetRead {
  endpoint: string;
  poll_interval_s: number | null;
}

export interface WidgetEdit {
  kind: string;
  endpoint: string;
  schema_ref: string | null;
}

export interface HudWidget {
  id: string;
  title: string;
  scope: WidgetScope;
  visible_when: string | null;
  render_hint: string;
  read: WidgetRead;
  edit: WidgetEdit | null;
  refresh_on: string[];
  stale_threshold_s: number | null;
  owner_module: string | null;
}

export interface WidgetSnapshot {
  id: string;
  status: WidgetStatus;
  data: unknown;
  error: string | null;
  stale: boolean;
  title: string | null;
  render_hint: string | null;
}

export interface AggregateResult {
  campaign_id: string;
  scene_id: string | null;
  generated_at: string;
  widgets: WidgetSnapshot[];
}

export interface OrderedWidgetPayload {
  id: string;
  visible: boolean;
  options: Record<string, unknown>;
}

export interface WidgetGroupPayload {
  title: string;
  widgets: string[];
}

export interface HudConfigPayload {
  density: string;
  position: string;
  ordered_widgets: OrderedWidgetPayload[];
  groups: WidgetGroupPayload[];
  pinned_extras: Record<string, string[]>;
}

const base = (campaignId: string) => `/api/campaigns/${encodeURIComponent(campaignId)}/hud`;

export const hudApi = {
  aggregate(
    campaignId: string,
    signal?: AbortSignal,
    sceneId?: string | null,
  ): Promise<AggregateResult> {
    const url = sceneId
      ? `${base(campaignId)}?scene_id=${encodeURIComponent(sceneId)}`
      : base(campaignId);
    return api.get<AggregateResult>(url, { signal });
  },
  widget(
    campaignId: string,
    widgetId: string,
    signal?: AbortSignal,
    sceneId?: string | null,
  ): Promise<WidgetSnapshot> {
    const url = sceneId
      ? `${base(campaignId)}/widgets/${encodeURIComponent(widgetId)}?scene_id=${encodeURIComponent(sceneId)}`
      : `${base(campaignId)}/widgets/${encodeURIComponent(widgetId)}`;
    return api.get<WidgetSnapshot>(url, { signal });
  },
  available(campaignId: string, signal?: AbortSignal): Promise<HudWidget[]> {
    return api.get<HudWidget[]>(`${base(campaignId)}/widgets/available`, { signal });
  },
  getConfig(campaignId: string, signal?: AbortSignal): Promise<HudConfigPayload> {
    return api.get<HudConfigPayload>(`${base(campaignId)}/config`, { signal });
  },
  putConfig(campaignId: string, config: HudConfigPayload): Promise<HudConfigPayload> {
    return api.put<HudConfigPayload>(`${base(campaignId)}/config`, config);
  },
  resetConfig(campaignId: string): Promise<HudConfigPayload> {
    return api.post<HudConfigPayload>(`${base(campaignId)}/config/reset`);
  },
};
